src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2414,21 +2412,39 @@ Status Session::receive_download_message(const DownloadMessage& message)
         return status;
     }
 
-    version_type server_version = m_progress.download.server_version;
+    // Start with the download server version from the last download message, since the changesets in the new
+    // download message must have a remote version greater than (or equal to for FLX) this value.
+    version_type last_remote_version = m_progress.download.server_version;
     version_type last_integrated_client_version = m_progress.download.last_integrated_client_version;
+    // If there are 2 or more changesets in this message, track whether or not the changesets all have the same
+    // download_server_version to help with determining if this is a single message server-initiated bootstrap or not.
+    bool same_remote_version = last_in_batch && message.changesets.size() > 1;
+    bool after_first_changeset = false;
+
     for (const RemoteChangeset& changeset : message.changesets) {
-        // Check that per-changeset server version is strictly increasing, except in FLX sync where the server
-        // version must be increasing, but can stay the same during bootstraps.
-        bool good_server_version = m_is_flx_sync_session ? (changeset.remote_version >= server_version)
-                                                         : (changeset.remote_version > server_version);
+        // Check that per-changeset server version is strictly increasing since the last download server version,
+        // except in FLX sync where the server version must be increasing, but can stay the same during bootstraps.
+        bool good_server_version = m_is_flx_sync_session ? (changeset.remote_version >= last_remote_version)
+                                                         : (changeset.remote_version > last_remote_version);
         // Each server version cannot be greater than the one in the header of the download message.
         good_server_version = good_server_version && (changeset.remote_version <= progress.download.server_version);
         if (!good_server_version) {
             return {ErrorCodes::SyncProtocolInvariantFailed,
                     util::format("Bad server version in changeset header (DOWNLOAD) (%1, %2, %3)",
-                                 changeset.remote_version, server_version, progress.download.server_version)};
+                                 changeset.remote_version, last_remote_version, progress.download.server_version)};
+        }
+        // Check to see if all the changesets in this LastInBatch=true message have the same remote_version
+        // If so, this is a server-initiated single message bootstrap
+        if (same_remote_version && after_first_changeset) {
```

Comment:
if you decide to go with this approach, then the following is enough:
`bool same_remote_version = last_in_batch && message.changesets.size() > 1 && message.changesets.front().remote_version == message.changesets.back().remote_version`